Add a preinstall hook to check for tool versions

julienw opened the pull request.

gregtatum commented on bin/pre-install.js:
```diff
@@ -0,0 +1,99 @@
+  // parseFloat will get us major.minor and ignore the possible patch version, which is
+  // enough for our needs because node versions very rarely (or never ?) have
+  // patch versions.
+  const nodeVersion = parseFloat(strNodeVersion);
```
This does not work for node `v7.10.1`, parseFloat turns this into `7.1` which is less than `7.8`

julienw replied:
Oh, such a beginner mistake. I'm ashamed.

gregtatum commented on bin/pre-install.js:
```diff
@@ -0,0 +1,99 @@
+// The "userAgent" is something like 'yarn/0.28.1 npm/? node/v8.4.0 linux x64'
```
We need the license pre-amble. Could you also write some an intro comment explaining why this file exists.

julienw replied:
done

gregtatum commented on bin/pre-install.js:
```diff
@@ -0,0 +1,99 @@
+      `This project expects at least Node version ${expectedNodeVersion}.`
+    );
+    console.error(
+      'You can use a tool like `nvm` to install and manage installed node versions.\n'
```
Could you provide a link here to `nvm`'s install instructions?

gregtatum commented on bin/pre-install.js:
```diff
@@ -0,0 +1,99 @@
+
+function displayYarnVersionExplanation(prefix = '') {
+  console.error(
+    `${prefix}Yarn versions 0.26 and 0.27 have a bug that makes them unfit with this project.`
```
s/makes/make/

julienw replied:
Are you sure ? Isn't `the bug` that `makes`, which is 3rd singular form ?

gregtatum replied:
Hmm... I was looking at "Yarn versions" but you're probably right with "bug".

julienw commented on bin/pre-install.js:
```diff
@@ -0,0 +1,116 @@
+// This function compares two string versions. This compares only major.minor
+// and disrespect any textual prevision (like pre or beta).
+function versionCompare(a, b) {
+  return a.localeCompare(b, undefined, { numeric: true });
```
I tried with various versions of node, it looks like it's supported for a long time.

julienw commented on bin/pre-install.js:
```diff
@@ -0,0 +1,116 @@
+  const circleConfig = fs.readFileSync('circle.yml', { encoding: 'utf8' });
+  const expectedNodeVersion = /version:\s+([\d.]+)/.exec(circleConfig)[1];
+  if (!expectedNodeVersion) {
+    throw new Error(`Couldn't extract the node version from circle.yml.`);
```
Added this throw; this should show up in the CI (when it runs `yarn install`) and warn us when circle.yml is modified in a way that this script is broken.

julienw commented on bin/pre-install.js:
```diff
@@ -0,0 +1,116 @@
+  return expectedNodeVersion;
+}
+
+function displayYarnVersionExplanation(prefix = '') {
```
I realized this breaks in node v4. What do you think ?

julienw commented on bin/pre-install.js:
```diff
@@ -0,0 +1,116 @@
+}
+
+const agents = userAgent.split(' ').reduce((agents, agent) => {
+  const [key, value] = agent.split('/');
```
I realize that `const` and destructuring assignments do not work in older nodes. What do you think ?

julienw commented:
@gregtatum I added a commit with the last changes, would you mind reviewing again ? :)
I also added questions about which node version this script should support.

julienw commented:
> Also, a request I would have had is to move up the imperative logic into a function so that it's clear how the control flow logic works. I did that hear as well since flow doesn't understand process.exit()

Ah, TBH I actually wondered whether I should do it, when I tried adding flow too :)
